snappy/security.go

```diff
@@ -20,94 +20,241 @@
-	"encoding/json"
+	"bufio"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
 	"strings"
 
-	"gopkg.in/yaml.v2"
-
 	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/helpers"
 	"github.com/ubuntu-core/snappy/logger"
 	"github.com/ubuntu-core/snappy/pkg"
+	"github.com/ubuntu-core/snappy/policy"
+	"github.com/ubuntu-core/snappy/release"
 )
 
-type apparmorJSONTemplate struct {
-	Template      string   `json:"template"`
-	PolicyGroups  []string `json:"policy_groups"`
-	PolicyVendor  string   `json:"policy_vendor"`
-	PolicyVersion float64  `json:"policy_version"`
+type errPolicyNotFound struct {
+	// type of policy, e.g. template or cap
+	PolType string
+	// apparmor or seccomp
+	PolKind *securityPolicyType
+	// name of the policy
+	PolName string
 }
 
-type securitySeccompOverride struct {
-	Template      string   `yaml:"security-template,omitempty"`
-	PolicyGroups  []string `yaml:"caps,omitempty"`
-	Syscalls      []string `yaml:"syscalls,omitempty"`
-	PolicyVendor  string   `yaml:"policy-vendor"`
-	PolicyVersion float64  `yaml:"policy-version"`
+func (e *errPolicyNotFound) Error() string {
+	return fmt.Sprintf("could not find specified %s: %s (%s)", e.PolType, e.PolName, e.PolKind)
 }
 
-const defaultTemplate = "default"
+var (
+	// Note: these are true for ubuntu-core but perhaps not other flavors
+	defaultTemplate     = "default"
+	defaultPolicyGroups = []string{"network-client"}
 
-var defaultPolicyGroups = []string{"network-client"}
+	// AppArmor cache dir
+	aaCacheDir = "/var/cache/apparmor"
 
-// TODO: autodetect, this won't work for personal
-const defaultPolicyVendor = "ubuntu-core"
-const defaultPolicyVersion = 15.04
+	errOriginNotFound     = errors.New("could not detect origin")
+	errPolicyTypeNotFound = errors.New("could not find specified policy type")
+	errInvalidAppID       = errors.New("invalid APP_ID")
+	errPolicyGen          = errors.New("errors found when generating policy")
 
-func (s *SecurityDefinitions) generateApparmorJSONContent() ([]byte, error) {
-	t := apparmorJSONTemplate{
-		Template:      s.SecurityTemplate,
-		PolicyGroups:  s.SecurityCaps,
-		PolicyVendor:  defaultPolicyVendor,
-		PolicyVersion: defaultPolicyVersion,
+	// snappyConfig is the default securityDefinition for a snappy
+	// config fragment
+	snappyConfig = &SecurityDefinitions{
+		SecurityCaps: []string{},
 	}
+)
+
+var (
+	lsbRelease        = "/etc/lsb-release"
+	runAppArmorParser = runAppArmorParserImpl
+)
+
+func runAppArmorParserImpl(argv ...string) ([]byte, error) {
+	cmd := exec.Command(argv[0], argv[1:]...)
+	return cmd.CombinedOutput()
+}
+
+// SecuritySeccompOverrideDefinition is used to override seccomp security
+// defaults
+type SecuritySeccompOverrideDefinition struct {
+	Syscalls []string `yaml:"syscalls,omitempty" json:"syscalls,omitempty"`
+}
+
+// SecurityAppArmorOverrideDefinition is used to override apparmor security
+// defaults
+type SecurityAppArmorOverrideDefinition struct {
```

Comment:
Nevermind.. I now notice this is inside Snappy, and that you're struggling to namespace it all in a sensible way. This is all begging to be inside a package. Perhaps even a subpackage of a package, so we can work out the struct namings for marshaling without polluting the whole namespace. I won't bother you about this right now, though. Feels fine as a follow up refactoring.